File: Python/112_Path_Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    res = False

    def hasPathSum(self, root, targetSum):
        """
        :type root: Optional[TreeNode]
        :type targetSum: int
        :rtype: bool
        """

        def dfs(root, targetSum, curSum):
            if root == None:    # base
                if targetSum == curSum:  # base
                    self.res = True
                return

            curSum += root.val
            dfs(root.left, targetSum, curSum)
            dfs(root.right, targetSum, curSum)

            return

        if root != None:
            dfs(root, targetSum, 0)
        
        return self.res

# algorithm: DFS
# how to notice to detect a combination  
# time O(N) N is num of node
# memory O(N) 

File: C++/112_PathSum.cpp
using namespace std;

 struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
}; 

class Solution {
public:
    bool calDum(TreeNode* root, int targetSum, int curSum) {
        if (root!=nullptr){
            if (root->left==nullptr && root->right==nullptr) {
                if (targetSum == curSum + root->val) return true;
            }
        
            curSum += root->val;        
            return (calDum(root->left,targetSum,curSum) || calDum(root->right,targetSum,curSum));
        }
        return false;
    }

    bool hasPathSum(TreeNode* root, int targetSum) {
        return calDum(root, targetSum, 0);
    }
};